src/adapter/tracer/ptrace.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    cell::RefCell, collections::BTreeMap, fmt::Debug, marker::PhantomData, num::NonZeroU32, rc::Rc,
+    sync::Arc, time::Duration,
+};
+
+use serde::{Deserialize, Serialize};
+
+use std::time::Instant;
+
+use trustfall::{
+    FieldValue,
+    provider::{
+        Adapter, AsVertex, ContextIterator, ContextOutcomeIterator, EdgeParameters, Eid,
+        ResolveEdgeInfo, ResolveInfo, VertexInfo, VertexIterator, Vid,
+    },
+};
+
+/// Records and stores operations performed by the adapter.
+///
+/// This struct is intended for use inside of a TracingAdapter.
+/// Operations must be recorded sequentially in chronological order.
+/// Recording out-of-order operations will lead to invalid state.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Tracer {
+    pub calls: BTreeMap<FunctionCall, Vec<Duration>>,
```

Comment:
I'm worried about how big these `Vec` allocations might become. Have you tried recording an `aws-sdk-ec2` execution across all lints and measuring how much memory that takes?

When performance-profiling `cargo-semver-checks`, we should expect to have `O(# of lints)` traces be live concurrently. The number of lints in `cargo-semver-checks` is growing exponentially, so that seems not particularly future-proof.

Do we need _all_ durations, or would some summary statistics suffice? For example, say we had `min, max, sum, count, p50, p75, p90, p95, p99`, where the percentiles use an approximate counting method with reasonable accuracy bounds. Are there ways that lints might be slow that wouldn't be detectable from those statistics, but would be detectable from the full `Vec<Duration>`?